torchx/schedulers/test/ray_scheduler_test.py

```diff
@@ -266,3 +253,106 @@ def test_submit_dryrun_constructs_actor_command(self) -> None:
             self.assertEqual(
                 job.actors[0].command, "dummy_entrypoint1 arg1 dummy1.py arg2"
             )
+
+    class RayClusterSetup:
+        _instance = None  # pyre-ignore[4]
+
+        def __new__(cls):  # pyre-ignore[3]
+            if cls._instance is None:
+                cls._instance = super(RayClusterSetup, cls).__new__(cls)
+                ray.shutdown()
```

Comment:
if this is only to make sure that there was no active ray execution, lets just remove this. 
Is it possible to start ray on custom port? - if yes, lets just start it on custom port

Reply:
@amogkam curious to hear your take on this. Right now I'm using a singleton class to make sure that a Ray cluster gets correctly started and torn down. Then each test which needs to use this cluster would issue a decrement counter and when it hits 0 then cluster starts getting torn down. Is this how you do things in your tests?

Reply:
@msaroufim just to confirm- if you start and shutdown a ray cluster for each test individually instead of this singleton pattern, you run into this issue, is that correct: https://github.com/ray-project/ray/issues/9546?

For all of our tests we use `pytest` and not `unittest` so we have not run into this problem.

Reply:
Here's the pytest fixture that we use for testing Ray Jobs.

```python
@pytest.fixture
def call_ray_start(request):
    parameter = getattr(
        request, "param", "ray start --head --num-cpus=1 --min-worker-port=0 "
        "--max-worker-port=0 --port 0")
    command_args = parameter.split(" ")
    out = ray._private.utils.decode(
        subprocess.check_output(command_args, stderr=subprocess.STDOUT))
    # Get the redis address from the output.
    redis_substring_prefix = "--address='"
    address_location = (
        out.find(redis_substring_prefix) + len(redis_substring_prefix))
    address = out[address_location:]
    address = address.split("'")[0]

    yield address

    # Disconnect from the Ray cluster.
    ray.shutdown()
    # Kill the Ray cluster.
    subprocess.check_call(["ray", "stop"])
```

Might be switching back to what it was before the singleton pattern, but using `subprocess` to call ray start and ray stop instead of `os.system`.

Reply:
Ok this is good to know, I think I'll keep the test as is for now since it works just fine and will revisit this

Reply:
Synced with @msaroufim about this offline- the problem is that the tests can run in parallel, so there could be race conditions if you start and stop the ray cluster in each test individually. Singleton pattern was used to have a global setup and teardown to avoid this issue.